### PR TITLE
[SEDONA-498] Add ST_BestSRID

### PR DESCRIPTION
**Pull request by prantogg**
## Did you read the Contributor Guide?

- Yes, I have read [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-498. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

This PR introduces `ST_BestSRID`, which estimates the most appropriate SRID for a given geometry, based on its spatial extent and location.

- For geometries in the Arctic or Antarctic regions, the Lambert Azimuthal Equal Area projection is used.
- For geometries that fit within a single UTM zone and do not cross the International Date Line, a corresponding UTM SRID is chosen.
- In cases where none of the above conditions are met, or for geometries that cross the International Date Line, the function defaults to the Mercator projection.

## How was this patch tested?

- Passes new and existing tests

## Did this PR include necessary documentation updates?

- Yes, I am adding a new API. I am using the [current SNAPSHOT version number](https://github.com/apache/sedona/blob/99239524f17389fc4ae9548ea88756f8ea538bb9/pom.xml#L29) in since `vX.Y.Z` format.

### CODE REVIEW

**Comment by jiayuasu** [comment]
@prantogg Please add Snowflake doc

File: docs/api/flink/Function.md [hunk 1]
```diff
@@ -476,6 +476,33 @@ Output:
...
+## ST_BestSRID
+
+Introduction: Returns the estimated most appropriate Spatial Reference Identifier (SRID) for a given geometry, based on its spatial extent and location.  It evaluates the geometry's bounding envelope and selects an SRID that optimally represents the geometry on the Earth's surface. The function prioritizes Universal Transverse Mercator (UTM), Lambert Azimuthal Equal Area (LAEA), or falls back to the Mercator projection.
+
+- For geometries in the Arctic or Antarctic regions, the Lambert Azimuthal Equal Area projection is used.
+- For geometries that fit within a single UTM zone and do not cross the International Date Line, a corresponding UTM SRID is chosen.
+- In cases where none of the above conditions are met, or for geometries that cross the International Date Line, the function defaults to the Mercator projection.
+
+!!!Warning
+`ST_BestSRID` is designed to estimate a suitable SRID from a set of approximately 125 EPSG codes and works best for geometries that fit within the UTM zones. It should not be solely relied upon to determine the most accurate SRID, especially for specialized or high-precision spatial requirements.
```
Comment (furqaankhan) [inline docs/api/flink/Function.md:488]: I think this wouldn't render correctly as it is missing a tab.

File: docs/api/snowflake/vector-data/Function.md [hunk 2]
```diff
@@ -320,6 +320,33 @@ SELECT ST_Azimuth(ST_POINT(0.0, 25.0), ST_POINT(0.0, 0.0))
...
+## ST_BestSRID
+
+Introduction: Returns the estimated most appropriate Spatial Reference Identifier (SRID) for a given geometry, based on its spatial extent and location.  It evaluates the geometry's bounding envelope and selects an SRID that optimally represents the geometry on the Earth's surface. The function prioritizes Universal Transverse Mercator (UTM), Lambert Azimuthal Equal Area (LAEA), or falls back to the Mercator projection.
+
+- For geometries in the Arctic or Antarctic regions, the Lambert Azimuthal Equal Area projection is used.
+- For geometries that fit within a single UTM zone and do not cross the International Date Line, a corresponding UTM SRID is chosen.
+- In cases where none of the above conditions are met, or for geometries that cross the International Date Line, the function defaults to the Mercator projection.
+
+!!!Warning
+`ST_BestSRID` is designed to estimate a suitable SRID from a set of approximately 125 EPSG codes and works best for geometries that fit within the UTM zones. It should not be solely relied upon to determine the most accurate SRID, especially for specialized or high-precision spatial requirements.
```
Comment (furqaankhan) [inline docs/api/snowflake/vector-data/Function.md:332]: Same thing here, tab missing.

File: docs/api/flink/Function.md (hunk 1 above)
Comment (prantogg) [reply docs/api/flink/Function.md]: Fixed. Good catch. Indentations don't seem to transfer when copy pasting. Thanks!

File: docs/api/sql/Function.md [hunk 3]
```diff
@@ -474,6 +474,33 @@ Output:
 3.141592653589793
 ```
 
+## ST_BestSRID
+
+Introduction: Returns the estimated most appropriate Spatial Reference Identifier (SRID) for a given geometry, based on its spatial extent and location.  It evaluates the geometry's bounding envelope and selects an SRID that optimally represents the geometry on the Earth's surface. The function prioritizes Universal Transverse Mercator (UTM), Lambert Azimuthal Equal Area (LAEA), or falls back to the Mercator projection.
+
+- For geometries in the Arctic or Antarctic regions, the Lambert Azimuthal Equal Area projection is used.
+- For geometries that fit within a single UTM zone and do not cross the International Date Line, a corresponding UTM SRID is chosen.
+- In cases where none of the above conditions are met, or for geometries that cross the International Date Line, the function defaults to the Mercator projection.
```
Comment (Kontinuation) [inline docs/api/sql/Function.md:483]: I've not found how the data line was handled in the implementation. Would you please point out where the date line was handled in `bestSRID` function?

**Comment by jiayuasu** [comment]
Snowflake functions of this PR are being tested in a separate PR: https://github.com/wherobots/sedona-snowflake-tester/pull/11

**Comment by jiayuasu** [comment]
Snowflake test failed: https://github.com/wherobots/sedona-snowflake-tester/actions/runs/7984544420/job/21801519880?pr=11

File: docs/api/sql/Function.md (hunk 3 above)
Comment (prantogg) [reply docs/api/sql/Function.md:483]: It does not handle geometries crossing the IDL appropriately. By the below line I mean - geometries crossing the IDL are mishandled and cause the determined SRID to default to Mercator projection SRID.
The mishandling happens with the `angularWidth` and `angularHeight` calculation; this causes the `xwidth` to be much larger that 6 degrees, which defaults the SRID to mercator.
> In cases where none of the above conditions are met, or for geometries that cross the International Date Line, the function defaults to the Mercator projection.

File: docs/api/flink/Function.md [hunk 4]
```diff
@@ -476,6 +476,34 @@ Output:
...
+- For Geometries that cross the IDL, `ST_BestSRID` defaults the SRID to Mercator. Currently, `ST_BestSRID` does not handle geometries crossing the IDL.
+
+!!!Warning
+    `ST_BestSRID` is designed to estimate a suitable SRID from a set of approximately 125 EPSG codes and works best for geometries that fit within the UTM zones. It should not be solely relied upon to determine the most accurate SRID, especially for specialized or high-precision spatial requirements.
+
+Format: `ST_BestSRID(geom: Geometry)`
+
+Since: `v1.6.0`
+
+Spark SQL Example:
```
Comment (jiayuasu) [inline docs/api/flink/Function.md:495]: 1. Please call it `SQL Example` in all flink, spark, snowflake docs.
2. Please say that this ST_BestSRID functions takes a WGS84 geometry as input and the geometry should follow lon/lat order.

**Review by jiayuasu — CHANGES_REQUESTED** [review]

**Review by jiayuasu — APPROVED** [review]

**Review by Kontinuation — APPROVED** [review]

**Pull request merged**